test/controllers/users_controller_test.rb

```diff
@@ -0,0 +1,116 @@
+require "test_helper"
+
+describe UsersController do
+  describe "index" do
+    it "can get the index page" do
+
+      # Act
+      get users_path
+
+      # Assert
+      must_respond_with :success
+    end
+  end
+
+  describe "show" do
+    it "can get the user detail page" do
+
+      # Act
+      get user_path(User.first.id)
+      # Assert
+      must_respond_with :success
+    end
+
+    it "shows a 404 if user isn't found" do
+      user_id = -1
+      # Act
+      get user_path(user_id)
+
+      # Assert
+      must_respond_with :not_found
+    end
+  end
+
+  describe "login_form" do
+    it "can get the login form page" do
+
+      # Act
+      get login_path
+
+      # Assert
+      must_respond_with :success
+    end
+  end
+
+  describe "login" do
+    before do
+      @user = User.first
+      @login_data = {
+        user: {
+          name: @user.name,
+        },
+      }
+    end
+
+    it "Will respond with redirect" do
+      post login_path(@login_data)
+
+      must_redirect_to root_path
+    end
+    it "will login a returning user" do
+      post login_path(@login_data)
+
+      expect(session[:user_id]).must_equal @user.id
+    end
+
+    it "will create an account for and login a new user" do
+      @login_data[:user][:name] = "new user"
+
```

Comment:
Good work catching the two different success cases here.

Is there a failure case? What if the user forgets to type in their name before logging in?